eclipse/src/saros/ui/wizards/AddProjectToSessionWizard.java

```diff
@@ -668,4 +687,106 @@ private void displaySaveDirtyEditorsDialog(final List<IEditorPart> dirtyEditors)
       for (IEditorPart editor : dirtyEditors) editor.doSave(new NullProgressMonitor());
     }
   }
+
+  private void storeCurrentProjectNameMapping(final JID jid) {
+    Map<String, String> currentProjectNameMapping = new HashMap<>();
+
+    for (final ProjectNegotiationData data : negotiation.getProjectNegotiationData()) {
+      final String projectID = data.getProjectID();
+      currentProjectNameMapping.put(
+          data.getProjectName(), namePage.getTargetProjectName(projectID));
+    }
+
+    mappingStorage.updateMapping(jid, currentProjectNameMapping);
+  }
+
+  private static final class ResourceMappingStorage {
+    private static final String RESOURCE_MAPPING_STORE_PREFIX = "resource.mapping.storage.";
+
+    private static final CharSequence DELIMITER = ":";
+
+    private static final CharSequence ESCAPE_CHARACTER = "\\";
+
+    private static final CharSequence ESCAPE_SEQUENCE =
+        ESCAPE_CHARACTER.toString() + DELIMITER.toString();
+
+    private static final Pattern SPLIT_PATTERN =
+        Pattern.compile(
+            "(?<!"
+                + Pattern.quote(ESCAPE_CHARACTER.toString())
+                + ")"
+                + Pattern.quote(DELIMITER.toString()));
+
+    private final IPreferenceStore store;
+
+    public ResourceMappingStorage(final IPreferenceStore store) {
+      this.store = store;
+    }
+
+    public Map<String, String> getMapping(final JID jid) {
+      final Map<String, String> result = new HashMap<>();
+
+      final String mapping = store.getString(RESOURCE_MAPPING_STORE_PREFIX + jid.getBase());
+
+      if (mapping.isEmpty()) return result;
+
+      final String[] names = SPLIT_PATTERN.split(mapping);
+
+      for (int i = 0; i < names.length; i += 2) {
+        if (names[i].isEmpty() || names[i + 1].isEmpty()) continue;
+
+        result.put(unescape(names[i]), unescape(names[i + 1]));
+      }
+
+      return result;
+    }
+
+    public void updateMapping(final JID jid, final Map<String, String> resourceNameMapping) {
+
+      final Map<String, String> lastResourceNameMapping = getMapping(jid);
+
+      final Set<String> resourceNameDestinations = new HashSet<>(resourceNameMapping.values());
+
+      /*
+       * Remove existing mappings to avoid following problem:
+       * 1. We add A->A, B->B, C->C
+       * 2. Afterwards  A->A, B->C
+       * 3. When we now lookup A,B,C we will get A->A, B->C, C->C which is an invalid mapping
+       *
+       * In other words, ensure that the mapping is injective.
+       */
+      for (final Iterator<Entry<String, String>> it = lastResourceNameMapping.entrySet().iterator();
```

Comment:
I would prefer using a while loop defining the iterator beforehand as it is easier to read / not as unexpected.
My assumption is that you used a for loop to help the garbage collector (by binding the variable to the loop scope). But this is a bit confusing to read.

Reply:
Just a personal choice. I do not think is any great benefit either using for or while. I do not even know when the stack is corrected, but I guess "it" is popped of after leaving the for loop.